File: shoope/628-maximum-product-of-three-numbers/628.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
from typing import List, Optional


class Solution:
    def maximumProduct(self, nums: List[int]) -> int:
        nums = sorted(nums)
        return max((nums[0] * nums[1] * nums[-1]),
                   (nums[-3] * nums[-2] * nums[-1]))


if __name__ == '__main__':
    sol = Solution()
    nums = [1, 2, 3]
    print(sol.maximumProduct(nums))
    nums = [1, 2, 3, 4]
    print(sol.maximumProduct(nums))
    nums = [-1, -2, -3]
    print(sol.maximumProduct(nums))

File: shoope/628-maximum-product-of-three-numbers/628.go
package main

import (
    "sort"
    "fmt"
)

func maximumProduct(nums []int) int {
    sort.Ints(nums)
    n := len(nums)
    return max(nums[0] * nums[1] * nums[n-1], nums[n-1] * nums[n-2] * nums[n-3])
}

func max(a, b int) int {
    if (a < b) {
        return b
    }
    return a
}

func main() {
    nums := []int{1,2,3}
    fmt.Println(maximumProduct(nums))
    nums = []int{1,2,3,4}
    fmt.Println(maximumProduct(nums))
    nums = []int{-1,-2,-3}
    fmt.Println(maximumProduct(nums))
}